client/src/test/java/org/eclipse/hono/client/impl/HonoClientImplTest.java

```diff
@@ -150,6 +153,97 @@ public void testConnectFailsWithClientErrorForSecurityException(final TestContex
         assertTrue(connectionFactory.awaitFailure());
     }
 
+    /**
+     * Verifies that the client fails with a ClientErrorException with status code 401
+     * if it cannot authenticate to the server because no suitable SASL mechanism was found.
+     * 
+     * @param ctx The vert.x test client.
+     */
+    @Test
+    public void testConnectFailsWithClientErrorForNoSASLMechanismException(final TestContext ctx) {
+
+        // GIVEN a client that is configured to connect
+        // to a peer that always throws a SaslSystemException (as if no credentials were given)
+        props.setReconnectAttempts(2);
+        connectionFactory = new DisconnectHandlerProvidingConnectionFactory(con)
+                .setExpectedFailingConnectionAttempts(1) // only one connection attempt expected here
+                .failWith(new SaslSystemException(true, "Could not find a suitable SASL mechanism for the remote peer using the available credentials."));
+        client = new HonoClientImpl(vertx, connectionFactory, props);
+        final ProtonClientOptions options = new ProtonClientOptions()
+                .setConnectTimeout(10)
+                .setReconnectAttempts(0)
+                .setReconnectInterval(50);
+
+        // WHEN the client tries to connect
+        client.connect(options).setHandler(ctx.asyncAssertFailure(t -> {
+            // THEN the connection attempt fails due do lack of authorization
+            ctx.assertEquals(HttpURLConnection.HTTP_UNAUTHORIZED, ((ServiceInvocationException) t).getErrorCode());
+            ctx.assertEquals("no suitable SASL mechanism found for authentication with server", t.getMessage());
+        }));
+        // and the client has indeed tried three times in total
+        assertTrue(connectionFactory.awaitFailure());
+    }
+
+    /**
+     * Verifies that the client fails with a ServerErrorException with status code 503
+     * if it cannot authenticate to the server because of a transient error.
+     *
+     * @param ctx The vert.x test client.
+     */
+    @Test
+    public void testConnectFailsWithClientErrorForImpermanentSaslSystemException(final TestContext ctx) {
```

Comment:
`testConnectFailsWithClientErrorForTransientSaslSystemException`

Reply:
I've changed that.